File: website/docs/skills.ts
import { code, note } from "../markup";
import type { Page } from "../types";

export const skills: Page = {
  path: "docs/skills/index.html",
  title: "Skills",
  description:
    "Keep your Agent Skills in one library, install them into every connected tool, and let a session write down a procedure it worked out \u2014 for you to approve before it reaches anything.",
  kind: "docs",
  toc: [
    { label: "Why a library", id: "why" },
    { label: "Where skills go", id: "locations" },
    { label: "The format", id: "format" },
    { label: "Installing", id: "installing" },
    { label: "What each state means", id: "states" },
    { label: "Adopting what you already have", id: "adopting" },
    { label: "Project skills", id: "projects" },
    { label: "Skills agents write", id: "learning" },
    { label: "Limits", id: "limits" },
  ],
  body: `
    <h2 id="why">Why a library</h2>
    <p>Claude Code, Codex, and pi all read the <a href="https://agentskills.io">Agent Skills</a> open format, and each reads it from its own folder. A skill you want in all of them gets copied by hand as many times, and the copies start drifting the first time you improve one of them.</p>
    <p>Synapse keeps one copy. You edit it in one place, install it into every connected tool, and the app tells you which copies have fallen behind.</p>

    <h2 id="locations">Where skills go</h2>
    <table>
      <thead><tr><th>Place</th><th>Path</th></tr></thead>
      <tbody>
        <tr><td>The Synapse library</td><td><code>&lt;data&gt;/skills/&lt;name&gt;/SKILL.md</code></td></tr>
        <tr><td>&hellip; for one project</td><td><code>&lt;data&gt;/skills/@&lt;project&gt;/&lt;name&gt;/SKILL.md</code></td></tr>
        <tr><td>Claude Code</td><td><code>~/.claude/skills/&lt;name&gt;/SKILL.md</code></td></tr>
        <tr><td>Codex</td><td><code>~/.agents/skills/&lt;name&gt;/SKILL.md</code></td></tr>
        <tr><td>pi</td><td><code>~/.pi/agent/skills/&lt;name&gt;/SKILL.md</code></td></tr>
        <tr><td>A project's own</td><td><code>&lt;project&gt;/.claude/skills/</code>, <code>.agents/skills/</code>, <code>.pi/agent/skills/</code></td></tr>
      </tbody>
    </table>
    ${note("Codex reads personal skills from the shared <code>~/.agents/skills</code> folder rather than from its own home. <code>~/.codex/skills</code> holds the set Codex ships with, and Synapse does not write there.")}

    <h2 id="format">The format</h2>
    <p>A skill is a directory with a <code>SKILL.md</code> at its root. The frontmatter needs a <code>name</code> that matches the directory and a <code>description</code> saying what the skill does and when to reach for it. Everything after the frontmatter is the instructions.</p>
    ${code("markdown", `---
name: release-checklist
description: Walk the release checklist for this project. Use when cutting a release, tagging a version, or when the user asks what still needs doing before shipping.
---

## Steps

1. Run the full test suite.
2. Bump the version.
3. Write the release notes.`)}
    <p>A skill can carry more than its <code>SKILL.md</code> — <code>scripts/</code>, <code>references/</code>, and <code>assets/</code> travel with it, and Synapse copies the whole directory. Agents load only the name and description until a task calls for the skill, so a long reference file costs nothing until it is needed.</p>
    ${note("A description containing a colon followed by a space has to be quoted, or YAML reads it as a nested key. Synapse says so by name rather than passing the raw parser error through.")}
    <p>Synapse ships one skill, <code>synapse-mesh</code>, covering how to actually run a team of agents. It lands in your library on first use and is an ordinary editable skill from then on.</p>

    <h2 id="installing">Installing</h2>
    ${code("shell", `synapse skill list                     # what is in the library
synapse skill create my-workflow       # start one from a template
synapse skill edit my-workflow         # open it in $EDITOR
synapse skill install                  # copy everything into every tool
synapse skill install my-workflow      # or just one
synapse skill install --tool claude    # or into just one tool
synapse skill status                   # where each one is
synapse skill remove my-workflow       # take it back out`)}
    <p>Editing a skill in the library leaves the installed copies behind, and <code>status</code> says so. Running <code>install</code> again brings them back in step. In the app, the <strong>Skills</strong> screen does the same thing with a button per skill and one that installs everything.</p>

    <h2 id="states">What each state means</h2>
    <table>
      <thead><tr><th>State</th><th>What it means</th><th>What Synapse will do</th></tr></thead>
      <tbody>
        <tr><td>not installed</td><td>The tool does not have this skill.</td><td>Install it.</td></tr>
        <tr><td>installed</td><td>The copy matches the library.</td><td>Nothing needed.</td></tr>
        <tr><td>update available</td><td>Synapse installed it, and the library has changed since.</td><td>Install again to sync.</td></tr>
        <tr><td>changed in place</td><td>Synapse installed it, and it has been edited inside the tool.</td><td>Refuse, unless you pass <code>--replace</code>.</td></tr>
        <tr><td>not ours</td><td>A skill of the same name that Synapse never wrote.</td><td>Refuse. It is yours, not Synapse's.</td></tr>
        <tr><td>not in the library</td><td>The tool has a skill Synapse does not know about.</td><td>Nothing, until you adopt it.</td></tr>
        <tr><td>waiting for review</td><td>An agent wrote it and you have not looked at it yet.</td><td>Nothing, until you approve it.</td></tr>
      </tbody>
    </table>
    <p>Synapse knows which copies are its own because it records the digest it wrote and the library digest it came from. Without that record, "the library moved on" and "somebody wrote this by hand" look identical, and the safe response to both would be to do nothing.</p>

    <h2 id="adopting">Adopting what you already have</h2>
    <p>If a tool already has a skill you want managed, adopt it. That copies it into the library and records the tool it came from as already having it, so the original stops reading as somebody else's and starts staying in step.</p>
    ${code("shell", `synapse skill adopt humanize --tool claude
synapse skill install humanize`)}
    <p>The app lists these under <strong>Already in your tools</strong> on the Skills screen, with an Adopt button beside each.</p>

    <h2 id="projects">Project skills</h2>
    <p>A procedure you work out in one repository is usually about that repository. A library where every skill is global costs every session on the machine to hold one project's checklist, so skills have <strong>shelves</strong>: the library root for what is true everywhere, and one shelf per project beside it.</p>
    ${code("shell", `synapse skill create release --project        # this repository's own
synapse skill install release --project      # into this repository's .claude/skills
synapse skill list --global                  # or just the shared ones`)}
    <p>A project skill installs into that project's own skills folders rather than into your home, so it is loaded when you work there and nowhere else. It also travels with the repository once it is installed, which is the point.</p>
    <p>Every command that takes a bare name looks at this project's shelf before the global one, so <code>synapse skill show release</code> run inside a repository shows that repository's version. A global skill and a project skill can share a name without colliding; <code>--global</code> and <code>--project [folder]</code> say which you mean when it matters.</p>
    ${note("Not every tool has a place for these", "A project skill needs a project-local skills folder, and a tool described without one simply has nowhere to put it. Synapse leaves that pairing out of <code>skill status</code> rather than reporting it as something to fix.")}

    <h2 id="learning">Skills agents write</h2>
    <p>Switch this on and a session can write down a procedure it worked out, and correct one that turned out wrong:</p>
    ${code("shell", `synapse settings learn on`)}
    <p>That adds two tools, <code>teach</code> and <code>revise</code>, to every connected session, and the guidance explaining them arrives with them. It is off until you ask for it, because a tool definition costs context in every session that loads it.</p>
    <p>Synapse has no model and runs no loop, so it never reflects on a session. The session decides what it learned; Synapse decides where that lands and who has to agree to it.</p>
    <h3>Nothing reaches a tool until you say so</h3>
    <p>A skill an agent writes is <strong>proposed</strong>. It is in the library and in no tool, and it stays there until you approve it. The gate is on installing rather than on writing, because the library is Synapse's own folder while a skill's description is loaded into every session of every tool holding it — so writing one costs you a line in a list, and installing it is the decision.</p>
    ${code("shell", `synapse skill proposed                  # what is waiting, and who wrote it
synapse skill show cut-a-release        # read it
synapse skill approve cut-a-release     # install it where it belongs
synapse skill reject cut-a-release --confirm`)}
    <p><code>synapse skill install</code> with no name steps over anything waiting, so a proposal never reaches a tool because you installed everything. The app shows the same queue on the <strong>Skills</strong> screen with Approve and Turn down beside each, and the terminal dashboard does it with <code>a</code> and <code>d</code>.</p>
    <p>You will also see it said once at the start of a session — <code>Synapse connected · 1 skill to review</code> — and nowhere else. Waiting is what a proposal is for; a queue that interrupts is not a queue.</p>
    <h3>Corrections do reach the copies</h3>
    <p><code>revise</code> is the deliberate exception. You already agreed to that skill being loaded, and a correction that never arrives leaves every session running the version that was wrong — so a revision goes out to the copies Synapse installed. It reaches only copies Synapse wrote that nobody has edited since; one you changed by hand is yours and is left where it is.</p>
    <p>Nothing is lost doing it. What the skill said before is kept, and putting it back is one command:</p>
    ${code("shell", `synapse skill history cut-a-release     # what it used to say, and what was wrong
synapse skill revert cut-a-release      # put the last version back`)}
    <p>Reverting is itself a revision, so a revert can be reverted. The newest twenty versions of each skill are kept.</p>

    <h2 id="limits">Limits</h2>
    <ul>
      <li>Installing copies the directory rather than linking it, so a copy stays put if the library moves. The cost is that an edit in the library has to be installed again to reach the tools, which is what <code>status</code> is for.</li>
      <li>Deleting a skill from the library does not remove the copies already installed. Use <code>synapse skill remove</code> for that, before or after.</li>
      <li>A skill whose <code>SKILL.md</code> does not parse is reported and skipped rather than copied anywhere.</li>
      <li>A skill an agent writes is only ever a skill. It cannot install itself, cannot reach a tool, and cannot write anywhere except the library.</li>
    </ul>
  `,
};

File: crates/synapsecore/src/cli/skills.rs
//! Managing the skill library and getting it into the tools that read it.
//!
//! Two things here are not obvious from the command names. A skill an agent
//! wrote is *proposed*: it is in the library and in no tool, and a bulk
//! `install` steps over it, because the gate on self-improvement is install and
//! not write. And every command that takes a bare name resolves this project's
//! shelf before the global one, so a repository's own `release` shadows the
//! shared one instead of colliding with it.

use crate::agent::Agent;
use crate::cli::Outcome;
use crate::cli::editor::editor;
use crate::skill::{self, Receipts, Shelf, Skill};
use anyhow::{Context, Result};
use std::ffi::OsString;
use std::path::PathBuf;

const USAGE: &str = "usage: synapse skill <list|show|create|edit|delete|install|remove|status|adopt|proposed|approve|reject|history|revert>";

pub fn run(arguments: &[OsString]) -> Result<Outcome> {
    let action = super::relay::text(arguments, 0, USAGE)?;
    let json = arguments.iter().any(|value| value == "--json");
    match action.as_str() {
        "list" => list(arguments, json)?,
        "show" => show(arguments)?,
        "create" => create(arguments)?,
        "edit" => edit(arguments)?,
        "delete" => delete(arguments)?,
        "install" => install(arguments)?,
        "remove" => remove(arguments)?,
        "status" => status(arguments, json)?,
        "adopt" => adopt(arguments)?,
        "proposed" => proposed(json)?,
        "approve" => approve(arguments)?,
        "reject" => reject(arguments)?,
        "history" => history(arguments, json)?,
        "revert" => revert(arguments)?,
        unknown => anyhow::bail!("unknown skill command `{unknown}`\n\n{USAGE}"),
    }
    Ok(Outcome::Exit(0))
}

fn list(arguments: &[OsString], json: bool) -> Result<()> {
    let (skills, problems) = skill::library::all()?;
    let runtime = tokio::runtime::Runtime::new()?;
    let receipts = runtime.block_on(Receipts::open(crate::files::database()?))?;
    let waiting = runtime.block_on(receipts.proposals())?;
    let wanted = shelf(arguments)?;

    let selected: Vec<&Skill> = skills
        .iter()
        .filter(|skill| wanted.as_ref().is_none_or(|shelf| &skill.shelf == shelf))
        .collect();

    if json {
        println!("{}", serde_json::to_string_pretty(&selected)?);
    } else {
        for skill in selected {
            let proposed = waiting
                .iter()
                .any(|item| item.shelf == skill.shelf.key() && item.skill == skill.name);
            println!(
                "{}\t{}\t{}\t{}{}",
                skill.name,
                skill.shelf.label(),
                skill.files.len(),
                if proposed { "proposed · " } else { "" },
                headline(&skill.description)
            );
        }
    }
    report(&problems);
    Ok(())
}

fn show(arguments: &[OsString]) -> Result<()> {
    let name = super::relay::text(arguments, 1, "usage: synapse skill show <name>")?;
    let skill = find(arguments, &name)?;
    let root = skill::library::path(&skill.shelf, &skill.name)?;
    print!("{}", std::fs::read_to_string(root.join(skill::ENTRY))?);
    Ok(())
}

fn create(arguments: &[OsString]) -> Result<()> {
    let name = super::relay::text(arguments, 1, "usage: synapse skill create <name>")?;
    // A skill somebody types out by hand is global unless they say otherwise.
    // The volume of project skills comes from agents, and they pass a scope.
    let shelf = shelf(arguments)?.unwrap_or(Shelf::Global);
    let root = skill::library::create(&shelf, &name)?;
    println!("Created {}", root.join(skill::ENTRY).display());
    println!("Edit it, then run `synapse skill install {name}`.");
    Ok(())
}

fn edit(arguments: &[OsString]) -> Result<()> {
    let name = super::relay::text(arguments, 1, "usage: synapse skill edit <name>")?;
    let skill = find(arguments, &name)?;
    let entry = skill::library::path(&skill.shelf, &skill.name)?.join(skill::ENTRY);
    let current = std::fs::read_to_string(&entry)
        .with_context(|| format!("no skill named `{name}` in the library"))?;
    let draft = std::env::temp_dir().join(format!("synapse-skill-{name}.md"));
    std::fs::write(&draft, &current)?;

    let status = std::process::Command::new(editor())
        .arg(&draft)
        .status()
        .context("could not open your editor")?;
    anyhow::ensure!(status.success(), "the editor exited without saving");

    let edited = std::fs::read_to_string(&draft)?;
    match skill::library::save(&skill.shelf, &name, &edited) {
        Ok(_) => {
            let _ = std::fs::remove_file(&draft);
            println!("Saved {}", entry.display());
            println!("Run `synapse skill install {name}` to push it to your tools.");
            Ok(())
        }
        Err(error) => {
            Err(error).with_context(|| format!("your draft is still at {}", draft.display()))
        }
    }
}

fn delete(arguments: &[OsString]) -> Result<()> {
    let name = super::relay::text(arguments, 1, "usage: synapse skill delete <name> --confirm")?;
    anyhow::ensure!(
        arguments.iter().any(|value| value == "--confirm"),
        "add --confirm to delete `{name}` from the library"
    );
    let skill = find(arguments, &name)?;
    let runtime = tokio::runtime::Runtime::new()?;
    let receipts = runtime.block_on(Receipts::open(crate::files::database()?))?;
    let root = skill::library::delete(&skill.shelf, &name)?;
    runtime.block_on(receipts.forgethistory(skill.shelf.key(), &name))?;
    println!("Deleted {}", root.display());
    println!("Copies already installed in your tools are left alone.");
    Ok(())
}

fn install(arguments: &[OsString]) -> Result<()> {
    let runtime = tokio::runtime::Runtime::new()?;
    let receipts = runtime.block_on(Receipts::open(crate::files::database()?))?;
    let agents = chosen(arguments)?;
    let replace = arguments.iter().any(|value| value == "--replace");
    let (skills, problems) = skill::library::all()?;
    let waiting = runtime.block_on(receipts.proposals())?;
    let wanted = named(arguments);

    let selected: Vec<&Skill> = match &wanted {
        Some(name) => {
            let skill = find(arguments, name)?;
            anyhow::ensure!(
                !waiting
                    .iter()
                    .any(|item| item.shelf == skill.shelf.key() && item.skill == skill.name),
                "`{name}` is waiting for review; approve it with `synapse skill approve {name}`"
            );
            skills
                .iter()
                .filter(|item| item.name == skill.name && item.shelf == skill.shelf)
                .collect()
        }
        // Everything means everything approved. A proposal reaching a tool
        // because somebody ran a bulk install is the one way this gate leaks.
        None => skills
            .iter()
            .filter(|skill| {
                !waiting
                    .iter()
                    .any(|item| item.shelf == skill.shelf.key() && item.skill == skill.name)
            })
            .collect(),
    };
    anyhow::ensure!(
        !selected.is_empty(),
        "no skill named `{}` in the library",
        wanted.unwrap_or_default()
    );

    let mut failures = Vec::new();
    for agent in &agents {
        for skill in &selected {
            if skill::target(agent, &skill.shelf, &skill.name).is_none() {
                continue;
            }
            match runtime.block_on(skill::install(&receipts, agent, skill, replace)) {
                Ok(_) => println!("{} → {}", skill.name, agent.name),
                Err(error) => failures.push(format!("{} → {}: {error:#}", skill.name, agent.name)),
            }
        }
    }
    report(&problems);
    for failure in &failures {
        eprintln!("warning: {failure}");
    }
    anyhow::ensure!(
        failures.is_empty(),
        "{} skill install(s) did not happen",
        failures.len()
    );
    Ok(())
}

fn remove(arguments: &[OsString]) -> Result<()> {
    let name = super::relay::text(
        arguments,
        1,
        "usage: synapse skill remove <name> [--tool <tool>]",
    )?;
    let runtime = tokio::runtime::Runtime::new()?;
    let receipts = runtime.block_on(Receipts::open(crate::files::database()?))?;
    let force = arguments.iter().any(|value| value == "--force");
    // Resolved from the library when it is still there, and from the global
    // shelf when it is not: a skill deleted from the library still has copies.
    let key = find(arguments, &name)
        .map(|skill| skill.shelf.key().to_owned())
        .unwrap_or_default();
    for agent in chosen(arguments)? {
        match runtime.block_on(skill::remove(&receipts, &agent, &key, &name, force)) {
            Ok(true) => println!("Removed {} from {}", name, agent.name),
            Ok(false) => println!("{} has no {}", agent.name, name),
            Err(error) => eprintln!("warning: {} → {}: {error:#}", name, agent.name),
        }
    }
    Ok(())
}

fn status(arguments: &[OsString], json: bool) -> Result<()> {
    let home = crate::files::home()?;
    let runtime = tokio::runtime::Runtime::new()?;
    let (statuses, problems) = runtime.block_on(skill::survey(&home))?;
    let (skills, _) = skill::library::all()?;
    let known: Vec<String> = skills.iter().map(|skill| skill.name.clone()).collect();

    if json {
        let extra: Vec<_> = crate::agent::agents(&home)
            .into_iter()
            .map(|agent| {
                serde_json::json!({
                    "tool": agent.name,
                    "unknown": skill::unknown(&agent, &known),
                })
            })
            .collect();
        println!(
            "{}",
            serde_json::to_string_pretty(&serde_json::json!({
                "skills": statuses,
                "unmanaged": extra,
            }))?
        );
        report(&problems);
        return Ok(());
    }

    let wanted = named(arguments);
    for entry in &statuses {
        if wanted.as_ref().is_some_and(|name| name != &entry.skill) {
            continue;
        }
        println!(
            "{}\t{}\t{}\t{}",
            entry.skill,
            entry.scope,
            entry.tool,
            match entry.proposed {
                true => "waiting for review",
                false => entry.state.label(),
            }
        );
    }
    for agent in crate::agent::agents(&home) {
        for name in skill::unknown(&agent, &known) {
            println!("{name}\tglobal\t{}\tnot in the library", agent.name);
        }
    }
    report(&problems);
    Ok(())
}

fn adopt(arguments: &[OsString]) -> Result<()> {
    let name = super::relay::text(
        arguments,
        1,
        "usage: synapse skill adopt <name> --tool <tool>",
    )?;
    let agents = chosen(arguments)?;
    let runtime = tokio::runtime::Runtime::new()?;
    let receipts = runtime.block_on(Receipts::open(crate::files::database()?))?;
    let mut adopted = false;
    for agent in &agents {
        match runtime.block_on(skill::adopt(&receipts, agent, &name)) {
            Ok(path) => {
                println!(
                    "Copied {} from {} into {}",
                    name,
                    agent.name,
                    path.display()
                );
                adopted = true;
                break;
            }
            Err(error) if agents.len() == 1 => return Err(error),
            Err(_) => continue,
        }
    }
    anyhow::ensure!(adopted, "no connected tool has a skill named `{name}`");
    println!("Run `synapse skill install {name}` to put it everywhere.");
    Ok(())
}

/// What agents have written and nobody has looked at. The queue is the whole
/// user-facing half of self-improvement: Synapse accumulates and waits.
fn proposed(json: bool) -> Result<()> {
    let runtime = tokio::runtime::Runtime::new()?;
    let receipts = runtime.block_on(Receipts::open(crate::files::database()?))?;
    let waiting = runtime.block_on(receipts.proposals())?;
    if json {
        println!("{}", serde_json::to_string_pretty(&waiting)?);
        return Ok(());
    }
    if waiting.is_empty() {
        println!("Nothing is waiting for review.");
        return Ok(());
    }
    for item in &waiting {
        println!(
            "{}\t{}\t{}\t{}",
            item.skill,
            item.scope,
            item.tool,
            headline(&item.note)
        );
        if !item.project.is_empty() {
            println!("\t{}", item.project);
        }
    }
    println!();
    println!("`synapse skill show <name>` to read one, then `approve` or `reject`.");
    Ok(())
}

fn approve(arguments: &[OsString]) -> Result<()> {
    let name = super::relay::text(arguments, 1, "usage: synapse skill approve <name>")?;
    let skill = find(arguments, &name)?;
    let agents = chosen(arguments)?;
    let replace = arguments.iter().any(|value| value == "--replace");
    let runtime = tokio::runtime::Runtime::new()?;
    let receipts = runtime.block_on(Receipts::open(crate::files::database()?))?;
    let results = runtime.block_on(skill::approve(&receipts, &agents, &skill, replace))?;
    anyhow::ensure!(
        !results.is_empty(),
        "no connected tool can hold a {} skill",
        skill.shelf.label()
    );
    for (tool, outcome) in results {
        match outcome {
            Ok(_) => println!("{name} → {tool}"),
            Err(error) => eprintln!("warning: {name} → {tool}: {error:#}"),
        }
    }
    Ok(())
}

fn reject(arguments: &[OsString]) -> Result<()> {
    let name = super::relay::text(arguments, 1, "usage: synapse skill reject <name> --confirm")?;
    anyhow::ensure!(
        arguments.iter().any(|value| value == "--confirm"),
        "add --confirm to turn down `{name}`"
    );
    let skill = find(arguments, &name)?;
    let runtime = tokio::runtime::Runtime::new()?;
    let receipts = runtime.block_on(Receipts::open(crate::files::database()?))?;
    let path = runtime.block_on(skill::reject(&receipts, &skill))?;
    println!("Removed {}", path.display());
    Ok(())
}

fn history(arguments: &[OsString], json: bool) -> Result<()> {
    let name = super::relay::text(arguments, 1, "usage: synapse skill history <name>")?;
    let skill = find(arguments, &name)?;
    let runtime = tokio::runtime::Runtime::new()?;
    let receipts = runtime.block_on(Receipts::open(crate::files::database()?))?;
    let history = runtime.block_on(receipts.revisions(skill.shelf.key(), &skill.name))?;
    if json {
        println!("{}", serde_json::to_string_pretty(&history)?);
        return Ok(());
    }
    if history.is_empty() {
        println!("`{name}` has never been revised.");
        return Ok(());
    }
    for revision in &history {
        println!(
            "{}\t{}\t{}",
            revision.id,
            match revision.tool.is_empty() {
                true => "you",
                false => &revision.tool,
            },
            headline(&revision.note)
        );
    }
    println!();
    println!("`synapse skill revert {name} [<id>]` puts one of these back.");
    Ok(())
}

fn revert(arguments: &[OsString]) -> Result<()> {
    let name = super::relay::text(arguments, 1, "usage: synapse skill revert <name> [<id>]")?;
    let skill = find(arguments, &name)?;
    let id = super::relay::text(arguments, 2, "")
        .ok()
        .filter(|value| !value.starts_with("--"))
        .map(|value| value.parse::<i64>())
        .transpose()
        .context("a revision id is a number from `synapse skill history`")?;
    let home = crate::files::home()?;
    let runtime = tokio::runtime::Runtime::new()?;
    let receipts = runtime.block_on(Receipts::open(crate::files::database()?))?;
    let (kept, reached) = runtime.block_on(skill::revert(&receipts, &home, &skill, id))?;
    println!("Reverted {name}; what it said is kept as revision {kept}.");
    match reached.is_empty() {
        true => println!("No tool was holding a copy Synapse could update."),
        false => println!("Updated in {}.", reached.join(", ")),
    }
    Ok(())
}

/// The shelf a command was told to use, or `None` when it should work it out
/// from the name.
fn shelf(arguments: &[OsString]) -> Result<Option<Shelf>> {
    if arguments.iter().any(|value| value == "--global") {
        return Ok(Some(Shelf::Global));
    }
    let mut found = arguments.iter().skip_while(|value| *value != "--project");
    let Some(_) = found.next() else {
        return Ok(None);
    };
    let root = found
        .next()
        .map(PathBuf::from)
        .filter(|value| !value.to_string_lossy().starts_with("--"))
        .map(Ok)
        .unwrap_or_else(std::env::current_dir)
        .context("could not determine the project folder")?;
    Ok(Some(Shelf::project(&root)))
}

/// The skill a bare name means: the shelf the caller named, or this project's
/// copy before the global one.
fn find(arguments: &[OsString], name: &str) -> Result<Skill> {
    match shelf(arguments)? {
        Some(shelf) => skill::library::read(&shelf, name),
        None => skill::library::locate(name, std::env::current_dir().ok().as_deref()),
    }
}

/// The positional name a command was given, if it was given one.
fn named(arguments: &[OsString]) -> Option<String> {
    super::relay::text(arguments, 1, "")
        .ok()
        .filter(|value| !value.starts_with("--"))
}

/// Which tools a command applies to. Everything by default, or one named with
/// `--tool`.
fn chosen(arguments: &[OsString]) -> Result<Vec<Agent>> {
    let home = crate::files::home()?;
    let agents = crate::agent::agents(&home);
    let Some(wanted) = super::relay::value(arguments, "--tool") else {
        return Ok(agents);
    };
    let wanted = wanted.to_lowercase();
    let known: Vec<String> = agents.iter().map(|agent| agent.slug.clone()).collect();
    let matched: Vec<Agent> = agents
        .into_iter()
        .filter(|agent| {
            agent.slug.eq_ignore_ascii_case(&wanted)
                || agent.command.eq_ignore_ascii_case(&wanted)
                || agent.name.to_lowercase() == wanted
                || agent.name.to_lowercase().replace(' ', "") == wanted.replace(' ', "")
        })
        .collect();
    anyhow::ensure!(
        !matched.is_empty(),
        "unknown tool `{wanted}`; this machine has {}",
        known.join(", ")
    );
    Ok(matched)
}

fn headline(description: &str) -> String {
    let line = description.split_whitespace().collect::<Vec<_>>().join(" ");
    match line.char_indices().nth(90) {
        Some((at, _)) => format!("{}…", &line[..at]),
        None => line,
    }
}

fn report(problems: &[String]) {
    for problem in problems {
        eprintln!("warning: skipped {problem}");
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn arguments(values: &[&str]) -> Vec<OsString> {
        values.iter().map(OsString::from).collect()
    }

    #[test]
    fn a_headline_is_one_trimmed_line() {
        assert_eq!(
            headline("Does a thing.\nAnd more."),
            "Does a thing. And more."
        );
        let long = headline(&"word ".repeat(50));
        assert!(long.ends_with('…'));
        assert!(long.chars().count() <= 91);
    }

    #[test]
    fn a_tool_can_be_named_by_command_or_label() {
        let selected = chosen(&arguments(&["install", "--tool", "claude"])).unwrap();
        assert_eq!(selected.len(), 1);
        assert_eq!(selected[0].command, "claude");

        assert_eq!(
            chosen(&arguments(&["install", "--tool", "Claude Code"]))
                .unwrap()
                .len(),
            1
        );
    }

    #[test]
    fn every_tool_is_the_default_and_an_unknown_one_is_refused() {
        assert_eq!(chosen(&arguments(&["install"])).unwrap().len(), 4);
        assert!(chosen(&arguments(&["install", "--tool", "emacs"])).is_err());
    }

    #[test]
    fn a_shelf_is_named_explicitly_or_worked_out_from_the_name() {
        assert!(shelf(&arguments(&["list"])).unwrap().is_none());
        assert_eq!(
            shelf(&arguments(&["list", "--global"])).unwrap(),
            Some(Shelf::Global)
        );

        let directory = tempfile::tempdir().unwrap();
        let named = shelf(&arguments(&[
            "create",
            "mine",
            "--project",
            &directory.path().display().to_string(),
        ]))
        .unwrap()
        .unwrap();
        assert_eq!(named.label(), "project");

        // `--project` with nothing after it means the folder you are in, not
        // the flag that follows it.
        let bare = shelf(&arguments(&["create", "mine", "--project", "--json"]))
            .unwrap()
            .unwrap();
        assert_eq!(
            bare.root().unwrap(),
            std::fs::canonicalize(std::env::current_dir().unwrap())
                .unwrap()
                .display()
                .to_string()
        );
    }

    #[test]
    fn a_positional_name_is_never_a_flag() {
        assert_eq!(named(&arguments(&["install", "mine"])), Some("mine".into()));
        assert_eq!(named(&arguments(&["install", "--json"])), None);
        assert_eq!(named(&arguments(&["install"])), None);
    }
}
